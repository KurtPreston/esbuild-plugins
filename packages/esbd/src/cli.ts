import { typecheckPlugin } from '@jgoz/esbuild-plugin-typecheck';
import { Command, Option } from 'commander';
import path from 'path';

import {
  BuildMode,
  CommandName,
  EsbdConfigWithPlugins,
  findConfigFile,
  readConfig,
} from './config';
import nodeDev from './esbd-node-dev';
import serve from './esbd-serve';
import { createLogger, Logger } from './log';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const { version } = require('../package.json');

interface ServeOptions {
  host?: string;
  port?: string;
  livereload?: boolean;
  servedir?: string;
  rewrite: boolean;
}

interface NodeDevOptions {
  respawn?: boolean;
}

function getEntryNameAndPath(entry: string): [entryPath: string, entryName: string | undefined] {
  const [entryName, entryPath] = entry.split(':');
  return entryPath ? [entryPath, entryName] : [entryName, undefined];
}

async function getConfigAndMode(
  commandName: CommandName,
  program: Command,
  entryPath: string,
  logger: Logger,
  watch?: boolean,
): Promise<[EsbdConfigWithPlugins, BuildMode]> {
  const absEntryPath = path.resolve(process.cwd(), entryPath);
  const maybeConfigPath = program.opts().config;
  const mode: BuildMode = program.opts().mode ?? 'development';

  const configPath = maybeConfigPath
    ? path.resolve(process.cwd(), maybeConfigPath)
    : await findConfigFile(path.dirname(absEntryPath));

  const config = (
    configPath ? await readConfig(path.resolve(process.cwd(), configPath), mode, commandName) : {}
  ) as EsbdConfigWithPlugins;

  config.outdir ??= program.opts().outdir;
  config.plugins ??= [];

  if (program.opts().check) {
    const buildMode = program.opts().tsBuildMode;
    config.plugins.push(
      typecheckPlugin({
        configFile: config.tsconfig,
        build: buildMode ? true : undefined,
        buildMode,
        logger,
        omitStartLog: true,
        watch,
      }),
    );
  }

  return [config, mode];
}

function commandWithGlobalOpts(program: Command, command: string) {
  return program
    .command(command)
    .option('-c, --config <path>', 'path to configuration file')
    .addOption(
      new Option('-m, --mode <mode>', 'output build mode').choices(['development', 'production']),
    )
    .option('-o, --outdir <path>', 'path to output directory')
    .option('-t, --check', 'check types asynchronously with the TypeScript compiler')
    .addOption(
      new Option('--ts-build-mode <mode>', 'TypeScript "build" mode behavior')
        .choices(['readonly', 'write-output'])
        .default('write-output'),
    );
}

export default function init() {
  const program = new Command('esbd').version(version);

  commandWithGlobalOpts(program, 'node-dev <entry>')
    .description('Node application development host')
    .option('-r, --respawn', 'restart program on exit/error (but quit after 3 restarts within 5s)')
    .addHelpText(
      'after',
      '\nArguments that appear after a special -- argument will be passed to the node program.' +
        '\n\nExample:' +
        '\n\tnode-dev path/to/entry.ts -- --port 8080 --config my-config.json',
    )
    .action(async (entry: string, options: NodeDevOptions, command: Command) => {
      const { respawn } = options;
      const logger = createLogger();
      const [entryPath, entryName] = getEntryNameAndPath(entry);
      const [config, mode] = await getConfigAndMode('node-dev', program, entryPath, logger, true);

      await nodeDev([entryPath, entryName], config, { args: command.args, logger, mode, respawn });
    });

  commandWithGlobalOpts(program, 'serve <entry>')
    .description('Single page application development server')
    .option('-d, --servedir <path>', 'directory of additional static assets to serve')
    .option('-l, --livereload', 'reload page on rebuild')
    .option('-h, --host <host>', 'IP/host name to use when serving requests', '0.0.0.0')
    .option('-p, --port <port>', 'port to use', '8000')
    .option('--rewrite', 'rewrite all not-found requests to "index.html" (SPA mode)', true)
    .option('--no-rewrite', 'disable request rewriting')
    .action(async (entry: string, options: ServeOptions) => {
      const { host = '0.0.0.0', port = '8000', livereload, servedir, rewrite } = options;
      const logger = createLogger();
      const [config, mode] = await getConfigAndMode('serve', program, entry, logger, true);

      await serve(entry, config, {
        mode,
        host,
        port: Number(port),
        livereload,
        logger,
        servedir: servedir ? path.resolve(process.cwd(), servedir) : undefined,
        rewrite,
      });
    });

  program.parse(process.argv);
}
