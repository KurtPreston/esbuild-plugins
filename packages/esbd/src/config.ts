import { BuildOptions, transform } from 'esbuild';
import fs from 'fs';
import K from 'kleur';
import { createRequire, Module } from 'module';
import path from 'path';
import vm from 'vm';

import type { HashAlgorithm } from './html-entry-point';

export type BuildMode = 'development' | 'production';

export interface EsbdConfig {
  /**
   * Base directory used for resolving entry points specified as relative paths.
   *
   * @default "process.cwd()"
   */
  basedir?: string;

  /**
   * Build options that will be merged with the generated esbuild config. Note that changing
   * some options will have no effect, such as "metafile" which always needs to be set
   * for esbd.
   */
  esbuild?: BuildOptions;

  /**
   * By default, assets (images, manifests, scripts, etc.) referenced by `<link>`, `<style>` and
   * `<script>` tags in the HTML template will be collected as esbuild assets if their `src` attributes
   * are specified as relative paths. The asset paths will be resolved relative to the *template file*
   * and will be copied to the output directory, taking `publicPath` into consideration if it has
   * been set.
   *
   * Absolute paths or URIs will be ignored.
   *
   * To ignore all `src` attributes and avoid collecting discovered assets, set this option to `true`.
   *
   * @default undefined
   */
  ignoreAssets?: boolean;

  /**
   * If specified, a cryptographic digest for each file referenced by a `<link>` or
   * `<script>` tag will be calculated using the specified algorithm and added as an
   * `integrity` attribute on the associated tag.
   * @see {@link https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity}
   *
   * @default undefined
   */
  integrity?: HashAlgorithm;
}

async function fileExists(configPath: string): Promise<boolean> {
  try {
    await fs.promises.access(configPath, fs.constants.R_OK);
    return true;
  } catch {
    return false;
  }
}

export async function findConfigFile(basedir: string): Promise<string | undefined> {
  const jsConfigPath = path.resolve(basedir, 'esbd.config.js');
  const tsConfigPath = path.resolve(basedir, 'esbd.config.ts');
  return (await fileExists(jsConfigPath))
    ? jsConfigPath
    : (await fileExists(tsConfigPath))
    ? tsConfigPath
    : undefined;
}

export async function readConfig(configPath: string, mode: BuildMode): Promise<EsbdConfig> {
  const loader = path.extname(configPath) === '.ts' ? 'ts' : undefined;
  const configSource = await fs.promises.readFile(configPath, 'utf-8');
  const configJs = await transform(configSource, { format: 'cjs', loader, target: 'node14' });

  if (configJs.warnings.length) {
    for (const warning of configJs.warnings) {
      console.warn(K.yellow(warning.text));
    }
  }

  const mainModule = require.main!;
  const contextModule = new Module(configPath, mainModule);
  contextModule.filename = configPath;
  contextModule.path = path.dirname(configPath);
  contextModule.paths = mainModule.paths;
  contextModule.require = createRequire(configPath);

  vm.runInNewContext(configJs.code, {
    __dirname: contextModule.path,
    __filename: contextModule.filename,
    exports: contextModule.exports,
    module: contextModule,
    require: contextModule.require,
    console: console,
  });

  contextModule.loaded = true;
  const exports = contextModule.exports;
  const output = exports.default ?? exports;

  let configObj: Record<string, any> | undefined;
  if (typeof output === 'function') {
    configObj = await output(mode);
  } else if (typeof output === 'object' && !!output) {
    configObj = output;
  }

  if (!configObj || typeof configObj !== 'object') {
    throw new Error(`Config file at "${configPath}" did not produce a valid configuration object.`);
  }

  const result: EsbdConfig = {
    basedir: configObj.basedir,
    esbuild: configObj.esbuild,
    ignoreAssets: configObj.ignoreAssets,
    integrity: configObj.integrity,
  };

  const resultKeys = new Set(Object.keys(result));
  const extraKeys = Object.keys(configObj).filter(key => !resultKeys.has(key));

  if (extraKeys.length) {
    console.warn(
      K.yellow(
        `Configuration has unknown keys "${extraKeys.join('", "')}". These will be ignored.`,
      ),
    );
  }

  return result;
}
