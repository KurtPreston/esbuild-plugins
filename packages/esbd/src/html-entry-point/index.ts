export * from './read-template';
export * from './types';
