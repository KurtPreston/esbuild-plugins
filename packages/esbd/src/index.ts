import bundle from './cli';

export type { BuildMode, CommandName, ConfigFn, EsbdConfig, EsbdConfigResult } from './config';

export { bundle, bundle as default };
