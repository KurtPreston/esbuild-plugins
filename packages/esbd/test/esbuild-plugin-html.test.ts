import { build, BuildOptions } from 'esbuild';
import fs from 'fs';
import path from 'path';
import prettier from 'prettier';

import { EntryPoints, EsbuildHtmlOptions, readTemplate } from '../src';

jest.mock('fs');

async function* walk(dirPath: string): AsyncIterable<string> {
  for await (const d of await fs.promises.readdir(dirPath, { withFileTypes: true })) {
    const entry = path.join(dirPath, d.name);
    if (d.isDirectory()) yield* walk(entry);
    else if (d.isFile()) yield entry;
  }
}

interface BuildWithHTMLOutput {
  html: string[];
  assets: string[];
}

const SEPARATOR = '---------------------------------';

expect.addSnapshotSerializer({
  test: value => typeof value === 'object' && !!value && 'html' in value && 'assets' in value,
  serialize: (val: unknown) => {
    const value = val as BuildWithHTMLOutput;
    const output: string[] = [];

    for (const html of value.html) {
      output.push(SEPARATOR);
      output.push(path.basename(html));
      output.push(SEPARATOR);

      const content = fs.readFileSync(html, 'utf-8');
      output.push(prettier.format(content, { parser: 'html' }));
    }

    for (const asset of value.assets) {
      output.push(SEPARATOR);
      output.push(path.basename(asset));
      output.push(SEPARATOR);

      const content = fs.readFileSync(asset, 'utf-8');
      output.push(content);
    }

    return output.join('\n');
  },
});

async function buildWithHTML(
  fixture: string | string[],
  options: Partial<EsbuildHtmlOptions> = {},
  buildOptionOverrides: Partial<BuildOptions> = {},
): Promise<BuildWithHTMLOutput> {
  const { absWorkingDir = __dirname, outdir = './out' } = buildOptionOverrides;
  const absOutDir = await fs.promises.mkdtemp(path.join(absWorkingDir, outdir + '-'));

  const fixtures = Array.isArray(fixture) ? fixture : [fixture];
  const html = await Promise.all(
    fixtures.map(f => readTemplate(`./fixture/${f}.html`, { basedir: __dirname, ...options })),
  );

  const entryPoints = html.reduce<EntryPoints>((obj, [entries]) => ({ ...obj, ...entries }), {});
  const buildOptions: BuildOptions = {
    absWorkingDir: __dirname,
    bundle: true,
    entryPoints,
    format: 'esm',
    splitting: true,
    write: false,
    plugins: html.map(([, plugin]) => plugin),
    ...buildOptionOverrides,
    outdir: path.relative(absWorkingDir, absOutDir),
  };

  // First build generates the actual output files, which is necessary because
  // we are using an in-memory FS that esbuild isn't aware of
  const { plugins, ...buildOptionsWithoutPlugins } = buildOptions;
  const result = await build(buildOptionsWithoutPlugins);
  for (const file of result.outputFiles!) {
    await fs.promises.writeFile(file.path, file.contents);
  }

  // Second build actually produces the HTML output
  await build(buildOptions);

  try {
    const html: string[] = [];
    const assets: string[] = [];

    for await (const entry of walk(absOutDir)) {
      if (entry.endsWith('.html')) html.push(entry);
      else assets.push(entry);
    }

    return { html, assets };
  } catch (e) {
    throw new Error(`Error reading output HTML: ${e}`);
  }
}

describe('eslint-plugin-html', () => {
  it('throws if outdir not set', () => {
    void expect(
      buildWithHTML('template-empty', {}, { logLevel: 'silent' }),
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"html-plugin: Unable to read template at ./fixture/template-empty.html"`,
    );
  });

  it('throws if template not found', () => {
    void expect(buildWithHTML('whoops', {}, { logLevel: 'silent' })).rejects.toThrowError(
      /Unable to read template at/,
    );
  });

  it('uses scripts and styles as entry points for the build', async () => {
    const output = await buildWithHTML('template-basic');
    expect(output).toMatchSnapshot();
  });

  it('adds integrity if requested', async () => {
    const output = await buildWithHTML('template-basic', { integrity: 'sha256' });
    expect(output).toMatchSnapshot();
  });

  it('copies and rebases assets from "href", "src" and "url" (with no public path)', async () => {
    const output = await buildWithHTML('template-assets');
    expect(output).toMatchSnapshot();
  });

  it('copies and rebases assets from "href", "src" and "url" (with FS public path)', async () => {
    const output = await buildWithHTML('template-assets', {}, { publicPath: '/static' });
    expect(output).toMatchSnapshot();
  });

  it('copies and rebases assets from "href", "src" and "url" (with URL public path)', async () => {
    const output = await buildWithHTML(
      'template-assets',
      {},
      { publicPath: 'http://test.com/static' },
    );
    expect(output).toMatchSnapshot();
  });

  it('skips asset rebasing/copying if "ignoreAssets" is set', async () => {
    const output = await buildWithHTML(
      'template-assets',
      { ignoreAssets: true },
      { publicPath: '/static' },
    );
    expect(output).toMatchSnapshot();
  });

  it('substitutes values from "define" in output HTML', async () => {
    const output = await buildWithHTML('template-define', {
      define: {
        'VERSION': '1.2.3',
        'process.env.NODE_ENV': 'development',
      },
    });
    expect(output).toMatchSnapshot();
  });

  it('can handle multiple entry points', async () => {
    const output = await buildWithHTML(
      'template-basic',
      {},
      {
        entryPoints: [
          './fixture/template-basic.js',
          './fixture/template-assets.css',
          './fixture/template-complex.js',
          './fixture/template-define.css',
        ],
      },
    );
    expect(output).toMatchSnapshot();
  });

  it('can be used on multiple templates', async () => {
    const output = await buildWithHTML(['template-basic', 'template-complex']);
    expect(output).toMatchSnapshot();
  });
});
