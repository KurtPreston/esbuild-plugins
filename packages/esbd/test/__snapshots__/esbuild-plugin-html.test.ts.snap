// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eslint-plugin-html adds integrity if requested 1`] = `
---------------------------------
template-basic.html
---------------------------------
<!DOCTYPE html>
<html>
  <head>
    <link
      href="z.css"
      rel="stylesheet"
      integrity="sha256-Q74SycgrBq0ejZ+ml+Itpcffb8CS70NRrTxZTJ027yU="
    />
    <script
      src="template-basic.js"
      type="module"
      integrity="sha256-QAamFdBResqMLdiEYs1XFcWxujo9AQQUdb++c8DCQcE="
    ></script>
  </head>
  <body></body>
</html>

---------------------------------
a-LTDY5RBP.js
---------------------------------
// fixture/deps/a.js
var A = "a";
export {
  A
};

---------------------------------
b-HIJ6EZP3.js
---------------------------------
// fixture/deps/b.js
var B = "b";
export {
  B
};

---------------------------------
c-RYJBTEBJ.js
---------------------------------
// fixture/deps/c.js
async function C() {
  const { A } = await import("./a-LTDY5RBP.js");
  const { B } = await import("./b-HIJ6EZP3.js");
  return A + B;
}
export {
  C
};

---------------------------------
template-basic.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

---------------------------------
template-basic.js
---------------------------------
// fixture/template-basic.js
import("./c-RYJBTEBJ.js").then(({ C }) => {
  return C();
}).then((output) => {
  console.log(output);
}).catch(() => {
  console.error("whoops");
});

---------------------------------
z.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

`;

exports[`eslint-plugin-html can be used on multiple templates 1`] = `
---------------------------------
template-basic.html
---------------------------------
<!DOCTYPE html>
<html>
  <head>
    <link href="z.css" rel="stylesheet" />
    <script src="template-basic.js" type="module"></script>
  </head>
  <body></body>
</html>

---------------------------------
template-complex.html
---------------------------------
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="description" content="Description" />
    <meta name="author" content="John Gozde" />
    <meta name="google" value="notranslate" />
    <link
      rel="apple-touch-icon"
      sizes="180x180"
      href="http://icons.com/icon.png"
    />
    <meta name="theme-color" content="#ffffff" />
    <meta
      name="viewport"
      content="width=device-width, user-scalable=no, initial-scale=1, maximum-scale=1, minimum-scale=1"
    />
    <title>Complex Template</title>
    <style type="text/css" media="screen, print"></style>
    <link rel="stylesheet" href="https://google.com/fonts" />
    <script>
      window.global = window;
    </script>
  </head>
  <body>
    <div class="app-container"></div>
    <script src="http://google.com/analytics"></script>
  </body>
</html>

---------------------------------
a-LTDY5RBP.js
---------------------------------
// fixture/deps/a.js
var A = "a";
export {
  A
};

---------------------------------
b-HIJ6EZP3.js
---------------------------------
// fixture/deps/b.js
var B = "b";
export {
  B
};

---------------------------------
c-RYJBTEBJ.js
---------------------------------
// fixture/deps/c.js
async function C() {
  const { A } = await import("./a-LTDY5RBP.js");
  const { B } = await import("./b-HIJ6EZP3.js");
  return A + B;
}
export {
  C
};

---------------------------------
template-basic.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

---------------------------------
template-basic.js
---------------------------------
// fixture/template-basic.js
import("./c-RYJBTEBJ.js").then(({ C }) => {
  return C();
}).then((output) => {
  console.log(output);
}).catch(() => {
  console.error("whoops");
});

---------------------------------
z.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

`;

exports[`eslint-plugin-html can handle multiple entry points 1`] = `
---------------------------------
template-basic.html
---------------------------------
<!DOCTYPE html>
<html>
  <head>
    <link href="./deps/z.css" rel="stylesheet" />
    <script src="template-basic.js" type="module"></script>
  </head>
  <body></body>
</html>

---------------------------------
a-LTDY5RBP.js
---------------------------------
// fixture/deps/a.js
var A = "a";
export {
  A
};

---------------------------------
b-HIJ6EZP3.js
---------------------------------
// fixture/deps/b.js
var B = "b";
export {
  B
};

---------------------------------
c-RYJBTEBJ.js
---------------------------------
// fixture/deps/c.js
async function C() {
  const { A } = await import("./a-LTDY5RBP.js");
  const { B } = await import("./b-HIJ6EZP3.js");
  return A + B;
}
export {
  C
};

---------------------------------
chunk-P5OBPOC4.js
---------------------------------

---------------------------------
template-assets.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

/* fixture/template-assets.css */
.assets {
  color: green;
}

---------------------------------
template-basic.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

---------------------------------
template-basic.js
---------------------------------
import "./chunk-P5OBPOC4.js";

// fixture/template-basic.js
import("./c-RYJBTEBJ.js").then(({ C }) => {
  return C();
}).then((output) => {
  console.log(output);
}).catch(() => {
  console.error("whoops");
});

---------------------------------
template-complex.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

---------------------------------
template-complex.js
---------------------------------
import "./chunk-P5OBPOC4.js";

// fixture/template-complex.js
import("./c-RYJBTEBJ.js").then(({ C }) => {
  return C();
}).then((output) => {
  console.log(output);
}).catch(() => {
  console.error("whoops");
});

---------------------------------
template-define.css
---------------------------------
/* fixture/template-define.css */
.define {
  color: green;
}

`;

exports[`eslint-plugin-html copies and rebases assets from "href", "src" and "url" (with FS public path) 1`] = `
---------------------------------
template-assets.html
---------------------------------
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="description" content="Description" />
    <meta name="author" content="John Gozde" />
    <meta name="google" value="notranslate" />
    <link rel="apple-touch-icon" sizes="180x180" href="/static/icon.png" />
    <link rel="mask-icon" href="/static/mask.svg" color="#5bbad5" />
    <meta name="theme-color" content="#ffffff" />
    <meta
      name="viewport"
      content="width=device-width, user-scalable=no, initial-scale=1, maximum-scale=1, minimum-scale=1"
    />
    <title>Complex Template</title>
    <style type="text/css" media="screen, print">
      @font-face {
        font-family: "Open Sans";
        font-weight: 400;
        font-style: normal;
        src: url("/static/font.svg#OpenSans") format("svg");
      }
      @font-face {
        font-family: "Open Sans";
        font-weight: 600;
        font-style: normal;
        src: url("/static/font.svg?OpenSans") format("svg");
      }
      body {
        background: url(/static/bg.png?test#foo);
        content: "url";
      }
    </style>
    <link rel="stylesheet" href="https://google.com/fonts" />
    <link rel="stylesheet" href="/static/custom.css" />
    <link href="/static/template-assets-style.css" rel="stylesheet" />
    <script>
      window.global = window;
    </script>
    <script src="/static/template-assets-script.js" type="module"></script>
  </head>
  <body>
    <div class="app-container"></div>
    <script src="http://google.com/analytics"></script>
  </body>
</html>

---------------------------------
a-LTDY5RBP.js
---------------------------------
// fixture/deps/a.js
var A = "a";
export {
  A
};

---------------------------------
b-HIJ6EZP3.js
---------------------------------
// fixture/deps/b.js
var B = "b";
export {
  B
};

---------------------------------
bg.png
---------------------------------
bg

---------------------------------
c-RYJBTEBJ.js
---------------------------------
// fixture/deps/c.js
async function C() {
  const { A } = await import("/static/a-LTDY5RBP.js");
  const { B } = await import("/static/b-HIJ6EZP3.js");
  return A + B;
}
export {
  C
};

---------------------------------
custom.css
---------------------------------
body {
  line-height: 1.4;
}

---------------------------------
font.svg
---------------------------------
<svg>font</svg>

---------------------------------
icon.png
---------------------------------
icon

---------------------------------
mask.svg
---------------------------------
<svg>mask</svg>

---------------------------------
template-assets-script.js
---------------------------------
// fixture/template-assets.js
import("/static/c-RYJBTEBJ.js").then(({ C }) => {
  return C();
}).then((output) => {
  console.log(output);
}).catch(() => {
  console.error("whoops");
});

---------------------------------
template-assets-style.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

/* fixture/template-assets.css */
.assets {
  color: green;
}

`;

exports[`eslint-plugin-html copies and rebases assets from "href", "src" and "url" (with URL public path) 1`] = `
---------------------------------
template-assets.html
---------------------------------
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="description" content="Description" />
    <meta name="author" content="John Gozde" />
    <meta name="google" value="notranslate" />
    <link
      rel="apple-touch-icon"
      sizes="180x180"
      href="http://test.com/icon.png"
    />
    <link rel="mask-icon" href="http://test.com/mask.svg" color="#5bbad5" />
    <meta name="theme-color" content="#ffffff" />
    <meta
      name="viewport"
      content="width=device-width, user-scalable=no, initial-scale=1, maximum-scale=1, minimum-scale=1"
    />
    <title>Complex Template</title>
    <style type="text/css" media="screen, print">
      @font-face {
        font-family: "Open Sans";
        font-weight: 400;
        font-style: normal;
        src: url("http://test.com/font.svg#OpenSans") format("svg");
      }
      @font-face {
        font-family: "Open Sans";
        font-weight: 600;
        font-style: normal;
        src: url("http://test.com/font.svg?OpenSans") format("svg");
      }
      body {
        background: url(http://test.com/bg.png?test#foo);
        content: "url";
      }
    </style>
    <link rel="stylesheet" href="https://google.com/fonts" />
    <link rel="stylesheet" href="http://test.com/custom.css" />
    <link
      href="http:/test.com/static/template-assets-style.css"
      rel="stylesheet"
    />
    <script>
      window.global = window;
    </script>
    <script
      src="http:/test.com/static/template-assets-script.js"
      type="module"
    ></script>
  </head>
  <body>
    <div class="app-container"></div>
    <script src="http://google.com/analytics"></script>
  </body>
</html>

---------------------------------
a-LTDY5RBP.js
---------------------------------
// fixture/deps/a.js
var A = "a";
export {
  A
};

---------------------------------
b-HIJ6EZP3.js
---------------------------------
// fixture/deps/b.js
var B = "b";
export {
  B
};

---------------------------------
bg.png
---------------------------------
bg

---------------------------------
c-RYJBTEBJ.js
---------------------------------
// fixture/deps/c.js
async function C() {
  const { A } = await import("http://test.com/static/a-LTDY5RBP.js");
  const { B } = await import("http://test.com/static/b-HIJ6EZP3.js");
  return A + B;
}
export {
  C
};

---------------------------------
custom.css
---------------------------------
body {
  line-height: 1.4;
}

---------------------------------
font.svg
---------------------------------
<svg>font</svg>

---------------------------------
icon.png
---------------------------------
icon

---------------------------------
mask.svg
---------------------------------
<svg>mask</svg>

---------------------------------
template-assets-script.js
---------------------------------
// fixture/template-assets.js
import("http://test.com/static/c-RYJBTEBJ.js").then(({ C }) => {
  return C();
}).then((output) => {
  console.log(output);
}).catch(() => {
  console.error("whoops");
});

---------------------------------
template-assets-style.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

/* fixture/template-assets.css */
.assets {
  color: green;
}

`;

exports[`eslint-plugin-html copies and rebases assets from "href", "src" and "url" (with no public path) 1`] = `
---------------------------------
template-assets.html
---------------------------------
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="description" content="Description" />
    <meta name="author" content="John Gozde" />
    <meta name="google" value="notranslate" />
    <link rel="apple-touch-icon" sizes="180x180" href="./assets/icon.png" />
    <link rel="mask-icon" href="./assets/mask.svg" color="#5bbad5" />
    <meta name="theme-color" content="#ffffff" />
    <meta
      name="viewport"
      content="width=device-width, user-scalable=no, initial-scale=1, maximum-scale=1, minimum-scale=1"
    />
    <title>Complex Template</title>
    <style type="text/css" media="screen, print">
      @font-face {
        font-family: "Open Sans";
        font-weight: 400;
        font-style: normal;
        src: url("font.svg#OpenSans") format("svg");
      }
      @font-face {
        font-family: "Open Sans";
        font-weight: 600;
        font-style: normal;
        src: url("font.svg?OpenSans") format("svg");
      }
      body {
        background: url(bg.png?test#foo);
        content: "url";
      }
    </style>
    <link rel="stylesheet" href="https://google.com/fonts" />
    <link rel="stylesheet" href="custom.css" />
    <link rel="stylesheet" href="./template-assets.css" />
    <script>
      window.global = window;
    </script>
  </head>
  <body>
    <div class="app-container"></div>
    <script src="http://google.com/analytics"></script>
    <script src="template-assets.js" type="module"></script>
  </body>
</html>

---------------------------------
a-LTDY5RBP.js
---------------------------------
// fixture/deps/a.js
var A = "a";
export {
  A
};

---------------------------------
b-HIJ6EZP3.js
---------------------------------
// fixture/deps/b.js
var B = "b";
export {
  B
};

---------------------------------
bg.png
---------------------------------
bg

---------------------------------
c-RYJBTEBJ.js
---------------------------------
// fixture/deps/c.js
async function C() {
  const { A } = await import("./a-LTDY5RBP.js");
  const { B } = await import("./b-HIJ6EZP3.js");
  return A + B;
}
export {
  C
};

---------------------------------
custom.css
---------------------------------
/* fixture/assets/custom.css */
body {
  line-height: 1.4;
}

---------------------------------
font.svg
---------------------------------
<svg>font</svg>

---------------------------------
template-assets.js
---------------------------------
// fixture/template-assets.js
import("./c-RYJBTEBJ.js").then(({ C }) => {
  return C();
}).then((output) => {
  console.log(output);
}).catch(() => {
  console.error("whoops");
});

`;

exports[`eslint-plugin-html skips asset rebasing/copying if "ignoreAssets" is set 1`] = `
---------------------------------
template-assets.html
---------------------------------
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="description" content="Description" />
    <meta name="author" content="John Gozde" />
    <meta name="google" value="notranslate" />
    <link rel="apple-touch-icon" sizes="180x180" href="./assets/icon.png" />
    <link rel="mask-icon" href="./assets/mask.svg" color="#5bbad5" />
    <meta name="theme-color" content="#ffffff" />
    <meta
      name="viewport"
      content="width=device-width, user-scalable=no, initial-scale=1, maximum-scale=1, minimum-scale=1"
    />
    <title>Complex Template</title>
    <style type="text/css" media="screen, print">
      @font-face {
        font-family: "Open Sans";
        font-weight: 400;
        font-style: normal;
        src: url("./assets/font.svg#OpenSans") format("svg");
      }
      @font-face {
        font-family: "Open Sans";
        font-weight: 600;
        font-style: normal;
        src: url("./assets/font.svg?OpenSans") format("svg");
      }
      body {
        background: url(./assets/bg.png?test#foo);
        content: "url";
      }
    </style>
    <link rel="stylesheet" href="https://google.com/fonts" />
    <link rel="stylesheet" href="./assets/custom.css" />
    <link href="/static/template-assets-style.css" rel="stylesheet" />
    <script>
      window.global = window;
    </script>
    <script src="/static/template-assets-script.js" type="module"></script>
  </head>
  <body>
    <div class="app-container"></div>
    <script src="http://google.com/analytics"></script>
  </body>
</html>

---------------------------------
a-LTDY5RBP.js
---------------------------------
// fixture/deps/a.js
var A = "a";
export {
  A
};

---------------------------------
b-HIJ6EZP3.js
---------------------------------
// fixture/deps/b.js
var B = "b";
export {
  B
};

---------------------------------
c-RYJBTEBJ.js
---------------------------------
// fixture/deps/c.js
async function C() {
  const { A } = await import("/static/a-LTDY5RBP.js");
  const { B } = await import("/static/b-HIJ6EZP3.js");
  return A + B;
}
export {
  C
};

---------------------------------
template-assets-script.js
---------------------------------
// fixture/template-assets.js
import("/static/c-RYJBTEBJ.js").then(({ C }) => {
  return C();
}).then((output) => {
  console.log(output);
}).catch(() => {
  console.error("whoops");
});

---------------------------------
template-assets-style.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

/* fixture/template-assets.css */
.assets {
  color: green;
}

`;

exports[`eslint-plugin-html substitutes values from "define" in output HTML 1`] = ``;

exports[`eslint-plugin-html uses scripts and styles as entry points for the build 2`] = `
---------------------------------
template-basic.html
---------------------------------
<!DOCTYPE html>
<html>
  <head>
    <link href="z.css" rel="stylesheet" />
    <script src="template-basic.js" type="module"></script>
  </head>
  <body></body>
</html>

---------------------------------
a-LTDY5RBP.js
---------------------------------
// fixture/deps/a.js
var A = "a";
export {
  A
};

---------------------------------
b-HIJ6EZP3.js
---------------------------------
// fixture/deps/b.js
var B = "b";
export {
  B
};

---------------------------------
c-RYJBTEBJ.js
---------------------------------
// fixture/deps/c.js
async function C() {
  const { A } = await import("./a-LTDY5RBP.js");
  const { B } = await import("./b-HIJ6EZP3.js");
  return A + B;
}
export {
  C
};

---------------------------------
template-basic.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

---------------------------------
template-basic.js
---------------------------------
// fixture/template-basic.js
import("./c-RYJBTEBJ.js").then(({ C }) => {
  return C();
}).then((output) => {
  console.log(output);
}).catch(() => {
  console.error("whoops");
});

---------------------------------
z.css
---------------------------------
/* fixture/deps/z.css */
.z {
  background-color: black;
}

`;
